pkg/controllers/clusterresourceplacement/controller.go

```diff
@@ -463,29 +523,76 @@ func (r *Reconciler) lookupLatestResourceSnapshot(ctx context.Context, crp *flee
+	latestSnapshot := &sortedList.Items[len(sortedList.Items)-1]
+	resourceIndex, err := parseResourceIndexFromLabel(latestSnapshot)
+	if err != nil {
+		klog.ErrorS(err, "Failed to parse the resource index label", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", klog.KObj(latestSnapshot))
+		return nil, -1, controller.NewUnexpectedBehaviorError(err)
+	}
+	return latestSnapshot, resourceIndex, nil
+}
+
+// listSortedResourceSnapshots returns the resource snapshots sorted by its index and its subindex.
+// The resourceSnapshot is less than the other one when resourceIndex is less.
+// When the resourceIndex is equal, then order by the subindex.
+// Note: the snapshot does not have subindex is the largest of a group and there should be only one in a group.
+func (r *Reconciler) listSortedResourceSnapshots(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement) (*fleetv1beta1.ClusterResourceSnapshotList, error) {
+	snapshotList := &fleetv1beta1.ClusterResourceSnapshotList{}
+	crpKObj := klog.KObj(crp)
+	if err := r.Client.List(ctx, snapshotList, client.MatchingLabels{fleetv1beta1.CRPTrackingLabel: crp.Name}); err != nil {
+		klog.ErrorS(err, "Failed to list all clusterResourceSnapshots", "clusterResourcePlacement", crpKObj)
+		return nil, controller.NewAPIServerError(false, err)
+	}
+	var errs []error
+	sort.Slice(snapshotList.Items, func(i, j int) bool {
+		ii, err := parseResourceIndexFromLabel(&snapshotList.Items[i])
 		if err != nil {
-			klog.ErrorS(err, "Failed to parse the resource index label", "clusterResourceSnapshot", klog.KObj(&snapshotList.Items[i]))
-			return nil, -1, controller.NewUnexpectedBehaviorError(err)
+			klog.ErrorS(err, "Failed to parse the resource index label", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", klog.KObj(&snapshotList.Items[i]))
+			errs = append(errs, err)
 		}
-		if lastResourceIndex < resourceIndex {
-			index = i
-			lastResourceIndex = resourceIndex
+		ji, err := parseResourceIndexFromLabel(&snapshotList.Items[j])
+		if err != nil {
+			klog.ErrorS(err, "Failed to parse the resource index label", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", klog.KObj(&snapshotList.Items[j]))
+			errs = append(errs, err)
 		}
+		if ii != ji {
+			return ii < ji
+		}
+
+		iDoesExist, iSubindex, err := utils.ExtractSubindexFromClusterResourceSnapshot(&snapshotList.Items[i])
+		if err != nil {
+			klog.ErrorS(err, "Failed to parse the subindex index", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", klog.KObj(&snapshotList.Items[j]))
+			errs = append(errs, err)
+		}
+		jDoesExist, jSubindex, err := utils.ExtractSubindexFromClusterResourceSnapshot(&snapshotList.Items[j])
+		if err != nil {
+			klog.ErrorS(err, "Failed to parse the subindex index", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", klog.KObj(&snapshotList.Items[j]))
+			errs = append(errs, err)
+		}
+		if !iDoesExist { // check if it's the first snapshot
```

Comment:
This lessFunc already compares the major index at an earlier step; so any two items can reach this point must have the same major index. If both of them do not have subindexes, it should be an error.